core/action/ControlPoint.cc

```diff
@@ -281,16 +283,40 @@ const std::vector<G4ThreeVector>& ControlPoint::GetFieldMask(const std::string&
             LOGSVC_WARN("Returning empty plan mask point vector");
         return m_plan_mask_points;
     } else if(type=="Sim") {
-        if(m_sim_mask_points.Get().empty())
+        if(m_cp_run.Get()->GetSimMaskPoints().empty())
             LOGSVC_WARN("Returning empty sim mask point vector");
-        return m_sim_mask_points.Get();
+        return m_cp_run.Get()->GetSimMaskPoints();
     }
     G4String msg = "Couldn't recognize control point field mask type!";
     LOGSVC_CRITICAL(msg.data());
     G4Exception("ControlPoint", "GetFieldMask", FatalErrorInArgument , msg);
     return m_plan_mask_points;
 }
 
+////////////////////////////////////////////////////////////////////////////////
+///
+void ControlPoint::FillSimFieldMask(const std::vector<G4PrimaryVertex*>& p_vrtx){
+
+    // Relative small number of iteration - lambda here will not hurt our performance?
+    auto getMaskPositioning = [&](G4PrimaryVertex* vrtx){
+    G4double x, y, zRatio = 0.;
+    G4double deltaX, deltaY, deltaZ;
+    G4ThreeVector position = vrtx->GetPosition();
+    deltaZ = 1000 - position.getZ();
+    zRatio = deltaZ / position.getZ(); 
+    x = position.getX() + zRatio * position.getX(); // x + deltaX;
+    y = position.getY() + zRatio * position.getY(); // y + deltaY;
+    return G4ThreeVector(x, y, 0);
```

Comment:
The question here... do we fixed the mask position at Z=0? In general, yes, but maybe we should put here the isocentre z value? Where the default value is 0;